File: algorithms/[872]叶子相似的树/solution.py
# topics = ["二叉树"]

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from __future__ import annotations
from typing import Generator, List, Optional


class TreeNode:
    def __init__(
        self,
        x: int = 0,
        left: Optional[TreeNode] = None,
        right: Optional[TreeNode] = None,
    ):
        self.val = x
        self.left = left
        self.right = right


class Solution:
    def leafSimilar(self, root1: TreeNode, root2: TreeNode) -> bool:
        """
        Binary Tree
        time O(n1 + n2), space O(n1 + n2), n1 和 n2 分别为两棵树的节点个数
        """
        return self.getLeaf(root1) == self.getLeaf(root2)

    def getLeaf(self, root: TreeNode) -> List[int]:
        res: List[int] = []

        def helper(node: TreeNode) -> None:
            if not node.left and not node.right:
                res.append(node.val)
                return
            if node.left:
                helper(node.left)
            if node.right:
                helper(node.right)

        helper(root)

        return res

    def leafSimilar2(self, root1: TreeNode, root2: TreeNode) -> bool:
        """
        生成器写法
        """

        def helper(node: TreeNode) -> Generator[int, None, None]:
            if not node.left and not node.right:
                yield node.val
            else:
                if node.left:
                    yield from helper(node.left)
                if node.right:
                    yield from helper(node.right)

        return list(helper(root1)) == list(helper(root2))

File: algorithms/[872]叶子相似的树/solution.ts
// topics = ["二叉树"]

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

export class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * Binary Tree & Generator
 * time O(n1 + n2), space O(n1 + n2), n1 和 n2 分别为两棵树的节点个数
 */
function leafSimilar(root1: TreeNode, root2: TreeNode): boolean {
  const dfs = function* (node: TreeNode): Generator<number, void, unknown> {
    if (!node.left && !node.right) {
      yield node.val;
    } else {
      if (node.left) {
        yield* dfs(node.left);
      }
      if (node.right) {
        yield* dfs(node.right);
      }
    }
  };

  const iter1 = dfs(root1);
  const iter2 = dfs(root2);
  let res1 = iter1.next();
  let res2 = iter2.next();

  while (res1.value == res2.value) {
    res1 = iter1.next();
    res2 = iter2.next();

    if (res1.done && res2.done) {
      return true;
    }
  }

  return false;
}
